File: 76. Minimum Window Substring/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) {
            return "";
        }

        int ansL = 0, ansR = -1, l = 0, need = 0, have = 0, minLen = s.length();
        
        Map<Character, Integer> needMap = new HashMap<>();
        for (int r = 0; r < t.length(); r++) {
            needMap.put(t.charAt(r), needMap.getOrDefault(t.charAt(r), 0) + 1);
            need = needMap.size();
        }

        Map<Character, Integer> haveMap = new HashMap<>();
        for (int r = 0; r < s.length(); r++) {
            if (have < need) {
                // if char is in need map
                    // increment char in have map
                    // check if have map's value and need map's value for char is equal
                        // if yes --> increment have
                if (needMap.containsKey(s.charAt(r))) {
                    int count = haveMap.getOrDefault(s.charAt(r), 0);
                    haveMap.put(s.charAt(r), count + 1);
                    if (count + 1 == needMap.get(s.charAt(r))) {
                        have++;
                    }
                }
            }
            // move left pointer
            while (have == need) {
                if (r - l + 1 <= minLen) {
                    ansR = r;
                    ansL = l;
                    minLen = r - l + 1;
                }

                Integer count = haveMap.get(s.charAt(l));
                if (count != null) {
                    haveMap.put(s.charAt(l), count - 1);
                    if (count - 1 < needMap.get(s.charAt(l))) {
                        have--;
                    }
                }
                l++;
            }
        }

        return s.substring(ansL, ansR + 1);
    }
}

File: 76. Minimum Window Substring/solution.go
package main

func minWindow(s, t string) string {
	if len(s) < len(t) {
		return ""
	}

	charCount := map[byte]int{}

	for right := 0; right < len(t); right++ {
		charCount[t[right]]++
	}

	start := 0
	minLength := len(s) + 1
	left := 0
	have := 0

	for right := 0; right < len(s); right++ {
		value, present := charCount[s[right]]
		if present {
			value--
			charCount[s[right]] = value
			if value == 0 {
				have++
			}
		}

		for have == len(charCount) {
			if right-left+1 < minLength {
				minLength = right - left + 1
				start = left
			}

			v, p := charCount[s[left]]
			if p {
				if v == 0 {
					have--
				}
				charCount[s[left]] = v + 1
			}
			left++
		}
	}

	if minLength > len(s) {
		return ""
	} else {
		return s[start : start+minLength]
	}
}
